### PR TITLE
Unpin scipy upper version

### LINKED ISSUE #862: Scipy 1.14.0 changed the upcasting rule of several special functions when passing integers
### Description

Functions like `scipy.special.psi` now always upcast to float64, even if the input is a low precision integer like `int8`. We need to handle these types, but:
1) float64 may be a bit too wasteful
2) We can't have the dtype depend on the scipy version, and lower pinning to 1.14.0 may be too restrictive?

### LINKED ISSUE #973: Simplify elemwise?
I also think of simplifying this part of code, it has a lot of redundant logic that is already managed by simple `np.asarray`

```diff
diff --git a/pytensor/tensor/elemwise.py b/pytensor/tensor/elemwise.py
index de966f1a7..a0892dfd0 100644
--- a/pytensor/tensor/elemwise.py
+++ b/pytensor/tensor/elemwise.py
@@ -767,31 +767,17 @@ class Elemwise(OpenMPOp):
         for i, (variable, storage, nout) in enumerate(
             zip(variables, output_storage, node.outputs)
         ):
-            if getattr(variable, "dtype", "") == "object":
-                # Since numpy 1.6, function created with numpy.frompyfunc
-                # always return an ndarray with dtype object
-                variable = np.asarray(variable, dtype=nout.dtype)
+
+            variable = np.asarray(variable, dtype=nout.dtype)

             if i in self.inplace_pattern:
                 odat = inputs[self.inplace_pattern[i]]
                 odat[...] = variable
                 storage[0] = odat

-            # Sometimes NumPy return a Python type.
-            # Some PyTensor op return a different dtype like floor, ceil,
-            # trunc, eq, ...
-            elif not isinstance(variable, np.ndarray) or variable.dtype != nout.dtype:
-                variable = np.asarray(variable, nout.dtype)
-                # The next line is needed for numpy 1.9. Otherwise
-                # there are tests that fail in DebugMode.
-                # Normally we would call pytensor.misc._asarray, but it
-                # is faster to inline the code. We know that the dtype
-                # are the same string, just different typenum.
-                if np.dtype(nout.dtype).num != variable.dtype.num:
-                    variable = variable.view(dtype=nout.dtype)
-                storage[0] = variable
+            storage[0] = variable
             # numpy.real return a view!
-            elif not variable.flags.owndata:
+            if not variable.flags.owndata:
                 storage[0] = variable.copy()
-            else:
-                storage[0] = variable
```

_Originally posted by @ferrine in https://github.com/pymc-devs/pytensor/issues/972#issuecomment-2286103831_

### PR DESCRIPTION
**Pull request by ferrine**
## Related Issue

- [ ] Closes #862
- [ ] Closes #973
- [ ] Related to #859 

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ferrine** [comment]
besides expected errors there is one that comes from mypy

```
!!!!!!!!!
2 files unexpectedly failed.
/home/runner/work/pytensor/pytensor/pytensor/gradient.py
/home/runner/work/pytensor/pytensor/pytensor/graph/basic.py
These files did not fail before, so please check the above output for errors in {PosixPath('/home/runner/work/pytensor/pytensor/pytensor/graph/basic.py'), PosixPath('/home/runner/work/pytensor/pytensor/pytensor/gradient.py')} and fix them.
You can run `python scripts/run_mypy.py --verbose` to reproduce this test locally.
```

File: pytensor/scalar/basic.py [hunk 1]
```diff
@@ -1142,12 +1142,15 @@ def output_types(self, types):
...
+                self.impl(*inputs),
+                dtype=node.outputs[0].dtype,
+            )
         else:
             variables = from_return_values(self.impl(*inputs))
             assert len(variables) == len(output_storage)
-            for storage, variable in zip(output_storage, variables):
-                storage[0] = variable
+            for out, storage, variable in zip(node.outputs, output_storage, variables):
+                storage[0] = np.asarray(variable, dtype=out.dtype)
```
Comment (ricardoV94) [inline pytensor/scalar/basic.py:1153]: Outputs should be scalars not arrays

Comment (ricardoV94) [reply pytensor/scalar/basic.py:1153]: Can't you just call `.astype(out.dtype)`?

Comment (ferrine) [reply pytensor/scalar/basic.py:1153]: They're scalar arrays, sometimes function returns floats and astype method is missing.

**Comment by ferrine** [comment]
I also think of simplifying this part of code, it has a lot of redundant logic that is already managed by simple `np.asarray`

```diff
diff --git a/pytensor/tensor/elemwise.py b/pytensor/tensor/elemwise.py
index de966f1a7..a0892dfd0 100644
--- a/pytensor/tensor/elemwise.py
+++ b/pytensor/tensor/elemwise.py
@@ -767,31 +767,17 @@ class Elemwise(OpenMPOp):
         for i, (variable, storage, nout) in enumerate(
             zip(variables, output_storage, node.outputs)
         ):
-            if getattr(variable, "dtype", "") == "object":
-                # Since numpy 1.6, function created with numpy.frompyfunc
-                # always return an ndarray with dtype object
-                variable = np.asarray(variable, dtype=nout.dtype)
+
+            variable = np.asarray(variable, dtype=nout.dtype)

             if i in self.inplace_pattern:
                 odat = inputs[self.inplace_pattern[i]]
                 odat[...] = variable
                 storage[0] = odat

-            # Sometimes NumPy return a Python type.
-            # Some PyTensor op return a different dtype like floor, ceil,
-            # trunc, eq, ...
-            elif not isinstance(variable, np.ndarray) or variable.dtype != nout.dtype:
-                variable = np.asarray(variable, nout.dtype)
-                # The next line is needed for numpy 1.9. Otherwise
-                # there are tests that fail in DebugMode.
-                # Normally we would call pytensor.misc._asarray, but it
-                # is faster to inline the code. We know that the dtype
-                # are the same string, just different typenum.
-                if np.dtype(nout.dtype).num != variable.dtype.num:
-                    variable = variable.view(dtype=nout.dtype)
-                storage[0] = variable
+            storage[0] = variable
             # numpy.real return a view!
-            elif not variable.flags.owndata:
+            if not variable.flags.owndata:
                 storage[0] = variable.copy()
             else:
                 storage[0] = variable
```

File: pytensor/scalar/basic.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/scalar/basic.py:1153]: Scalar arrays are not the same. These functions should return non array scalars

**Comment by ferrine** [comment]
@ricardoV94 how are these tests are supposed to pass?

https://github.com/pymc-devs/pytensor/blob/4d0103bd01af9eba3643403f68325e633e9304e5/tests/scalar/test_loop.py#L198

Why did they pass before?

File: pytensor/scalar/basic.py (hunk 1 above)
Comment (ferrine) [reply pytensor/scalar/basic.py:1153]: What are those?

Comment (ricardoV94) [reply pytensor/scalar/basic.py:1153]: `np.float64(5.0), np int32(3)` and so on

File: pytensor/scalar/basic.py [hunk 2]
```diff
@@ -1140,14 +1140,23 @@ def output_types(self, types):
         else:
             raise NotImplementedError(f"Cannot calculate the output types for {self}")
 
+    @staticmethod
+    def _cast_scalar(x, dtype):
+        if hasattr(x, "astype"):
+            return x.astype(dtype)
+        else:
+            return x
```
Comment (ricardoV94) [inline pytensor/scalar/basic.py:1148]: Does something like this work?
```suggestion
            return getattr(np, dtype)(x)
```

Comment (ferrine) [reply pytensor/scalar/basic.py:1148]: interesting

**Comment by ferrine** [comment]
Wow, looking into the broken test closer, it should have been a failing test!

before the change

image

after the change the dtype of `out` is float16 as declared in the graph and we see overflow

File: environment-osx-arm64.yml [hunk 3]
```diff
@@ -10,7 +10,7 @@ dependencies:
   - python=>3.10
   - compilers
   - numpy>=1.17.0,<2
-  - scipy>=0.14,<1.14.0
+  - scipy>=0.14.0
```
Comment (ricardoV94) [inline environment-osx-arm64.yml:13]: Can probably pin to the current major version. No way 0.x would still work anyway?

```suggestion
  - scipy>=1,<2
```

File: tests/scalar/test_loop.py [hunk 4]
```diff
@@ -212,12 +212,11 @@ def test_inner_composite(mode):
...
     assert y16.type.dtype == "float16"
 
-    fn32 = function([n_steps, x16], y16, mode=mode)
-    np.testing.assert_allclose(
-        fn32(n_steps=9, x16=np.array(4.73, dtype="float16")),
-        4.73 + 9,
-        rtol=1e-3,
-    )
+    fn16 = function([n_steps, x16], y16, mode=mode)
+    out16 = fn16(n_steps=9, x16=np.array(4.73, dtype="float16"))
```
[the remark this answers is not in the document]
Comment (ferrine) [reply tests/scalar/test_loop.py]: done

**Review by ricardoV94 — APPROVED** [review]
Thanks @ferrine

**Comment by ricardoV94** [comment]
What's with mypy? I guess it's not related, but can you fix it?

**Comment by ricardoV94** [comment]
@ferrine can you clean the commits into just 2:
1) scipy stuff 
2) mypy stuff
?

**Comment by ferrine** [comment]
Yes, will do soon

**Pull request merged**